php/Terminus/Tunnels.php

```diff
@@ -0,0 +1,182 @@
+<?php
+namespace Terminus;
+
+use Terminus\SiteFactory;
+
+/**
+ * @file
+ * Objects for Terminus tunnel support.
+ */
+
+
+class Tunnels {
+  /**
+   * Helper function to shell out, return status code, handle errors in a 
+   * regular way. 
+   * @param $cmd (string) the shell command to run.
+   * @param $lines (array) pass-by-reference variable for getting the output back out.
+   */
+  private function shellout($cmd, &$lines=array()) {
+    if (\Terminus::get_config('verbose')) {
+      \Terminus::log(sprintf('Running "%s"', $cmd));
+    }
+    exec($cmd, $lines, $status);
+    if ($status !== 0) {
+      \Terminus::error(sprintf('Tunnel command "%s" failed with exit code %d', $cmd, $return_var));
+      \Terminus::error(implode("\n", $lines));
+    }
+    return $status;
+  }
+
+  /**
+   * Helper function to open a new process.
+   * Inspired by Drush's drush_shell_proc_open()
+   * https://github.com/drush-ops/drush/blob/master/includes/exec.inc
+   * @param $cmd (string) the shell command to run.
+   */
+  private function proc_open($cmd) {
+    if (\Terminus::get_config('verbose')) {
+      \Terminus::log(sprintf('Opening process "%s"', $cmd));
+    }
+    $process = proc_open($cmd, array(0 => STDIN, 1 => STDOUT, 2 => STDERR), $pipes);
+    $proc_status = proc_get_status($process);
+    $exit_code = proc_close($process);
+    return ($proc_status["running"] ? $exit_code : $proc_status["exitcode"] );
+  }
+
+  /**
+   * Returns an array of open tunnels.
+   */
+  public function get_all() {
+    $cmd = 'ps -fU $USER |grep "ssh -[f]" | awk \'{print $2, $5, $12, $15}\'';
+    $status = $this->shellout($cmd, $lines);
+    if ($status == 0) {
+      $tunnels = array();
+      $sites = new SiteFactory();
+      $keys = array('raw', 'pid', 'created', 'port', 'host', 'type', 'environment', 'site_uuid');
+      // Tunnels are keyed by port, ports are keyed by env.site for mapping.
+      foreach ($lines as $line) {
+        $matches = array();
+        if (preg_match('/([0-9]{1,8}) ([0-9]{1,2}:[0-9]{2}PM|AM) ([0-9]{1,7}):127.0.0.1:[0-9]{1,7} [A-z0-9_]{3,16}\.[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}@(([A-z0-9_]{3,16})\.([A-z0-9_]{3,16})\.([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12})\.drush.in)/', $line, $matches)) {
+          $tunnel = array_combine($keys, $matches);
+          if (isset($data['site_uuid'])) {
+            $site = $sites->getSiteByUUID($tunnel['site_uuid']);
```

Comment:
It looks like nothing is being done with the site object itself. If it's not necessary, can we remove this logic?